strawberryfields/backends/tfbackend/ops.py

```diff
@@ -307,92 +317,96 @@ def loss_superop(T, D, batched=False):
+def single_displacement_matrix(r, phi, D, dtype=def_type.as_numpy_dtype):
+    """creates a single mode displacement matrix"""
+    r = r.numpy()
+    phi = phi.numpy()
+    gate = displacement_tw(r, phi, D, dtype)
+
+    def grad(dy):
+        Dr, Dphi = grad_displacement_tw(gate, r, phi)
+        grad_r = tf.math.real(tf.reduce_sum(dy * tf.math.conj(Dr)))
+        grad_phi = tf.math.real(tf.reduce_sum(dy * tf.math.conj(Dphi)))
+        return grad_r, grad_phi, None
+
+    return gate, grad
+
+
+def displacement_matrix(r, phi, D, batched=False):
+    """creates a single mode displacement matrix accounting for batching"""
+    r = tf.cast(r, def_type)
+    phi = tf.cast(phi, def_type)
     if batched:
-        idxs = batchify_indices(idxs, batch_size)
-        signs = signs * batch_size
-    sign_lower_diag = tf.cast(tf.SparseTensor(idxs, signs, dense_shape), tf.float32)
-    sign_matrix = tf.sparse.add(eye_diag, sign_lower_diag)
-    sign_matrix = tf.cast(tf.sparse.to_dense(sign_matrix), def_type)
-    lower_diag = tf.scatter_nd(idxs, tf.reshape(values, [-1]), dense_shape)
-    E = tf.cast(tf.eye(D), def_type) + lower_diag
-    E_prime = tf.math.conj(E) * sign_matrix
+        return tf.stack(
+            [single_displacement_matrix(r_, phi_, D) for r_, phi_ in tf.transpose([r, phi])]
+        )
+    return single_displacement_matrix(r, phi, D)
+
+
+@tf.custom_gradient
+def single_beamsplitter_matrix(theta, phi, D, dtype=def_type.as_numpy_dtype):
+    """creates a single mode beamsplitter matrix"""
+    theta = theta.numpy()
+    phi = phi.numpy()
+
+    gate = beamsplitter_tw(theta, phi, D, dtype)
+    gate = np.transpose(gate, [0, 2, 1, 3])
+
+    def grad(dy):
+        Dtheta, Dphi = grad_beamsplitter_tw(gate, theta, phi)
+        Dtheta = np.transpose(Dtheta, [0, 2, 1, 3])
+        Dphi = np.transpose(Dphi, [0, 2, 1, 3])
+        grad_theta = tf.math.real(tf.reduce_sum(dy * tf.math.conj(Dtheta)))
+        grad_phi = tf.math.real(tf.reduce_sum(dy * tf.math.conj(Dphi)))
+        return grad_theta, grad_phi, None
+
+    return gate, grad
+
+
+def beamsplitter_matrix(theta, phi, D, batched=False):
+    """creates a single mode beamsplitter matrix accounting for batching"""
+    theta = tf.cast(theta, def_type)
+    phi = tf.cast(phi, def_type)
     if batched:
-        eqn = "aik,ajk->aij"  # pylint: disable=bad-whitespace
-    else:
-        eqn = "ik,jk->ij"  # pylint: disable=bad-whitespace
-    prefactor = tf.expand_dims(
-        tf.expand_dims(tf.cast(tf.exp(-0.5 * tf.abs(alpha) ** 2), def_type), -1), -1
-    )
-    D_alpha = prefactor * tf.einsum(eqn, E, E_prime)
-    return D_alpha
+        return tf.stack(
+            [
+                single_beamsplitter_matrix(theta_, phi_, D)
+                for theta_, phi_ in tf.transpose([theta, phi])
+            ]
+        )
+    return tf.convert_to_tensor(single_beamsplitter_matrix(theta, phi, D))
 
 
-def beamsplitter_matrix(t, r, D, batched=False, save=False, directory=None):
```

Comment:
it's so good seeing all this going away. Bye Bye beamsplitter that requires `cutoff**5 ` in memory! Hello beamsplitter that runs in `cutoff**3` time.